src/zenml/cli/pipeline.py

```diff
@@ -138,3 +144,46 @@ def list_pipeline_runs(
             return
 
         cli_utils.print_pipeline_runs_table(pipeline_runs=pipeline_runs)
+
+
+@runs.command("delete")
+@click.argument("run_name_or_id", type=str, required=True)
+def delete_pipeline_run(run_name_or_id: str) -> None:
+    """Delete a pipeline run.
+
+    Args:
+        run_name_or_id: The name or ID of the pipeline run to delete.
+    """
+    cli_utils.print_active_config()
+    confirmation = cli_utils.confirmation(
+        f"Are you sure you want to delete pipeline run " f"`{run_name_or_id}`?"
+    )
+    if not confirmation:
+        cli_utils.declare("Pipeline run deletion canceled.")
+        return
+
+    client = Client()
+    run = client.get_pipeline_run(name_id_or_prefix=run_name_or_id)
+    run_artifacts: List["ArtifactResponseModel"] = []
+    for step_run in client.list_run_steps(pipeline_run_id=run.id):
+        if step_run.status == ExecutionStatus.COMPLETED:
+            run_artifacts.extend(step_run.output_artifacts.values())
+    if run_artifacts:
+        confirmation = cli_utils.confirmation(
+            f"Pipeline run '{run_name_or_id}' has produced "
+            f"{len(run_artifacts)} artifacts. Do you want to delete these "
+            "artifacts as well?"
+        )
+        if confirmation:
+            for artifact in run_artifacts:
+                client.delete_artifact(artifact_id=artifact.id)
```

Comment:
When you delete these artifacts from a completed step, what happens to all step runs that used the same artifact due to caching? Are they pointing to some non-existent entry now?

Reply:
However, we have a separate table `step_run_parents` which we could still use to connect nodes in the DAG even if there is no artifact between them.

Reply:
Mhm do you think it would make sense to instead insert something in the database that signals a deleted artifact?

The problem I see is that the thing we display in the UI aren't necessarily the output artifacts of a step, but the `outputs` in general. That step in the pipeline run still had those outputs, and to suddenly see my `trainer_step` without an output in the UI even though it returned a model before seems very confusing

Reply:
I now made it so deleting artifacts that are still used in any pipeline runs is not allowed.